### ISSUE TITLE
deps: go.mod announces compatibility with gogoproto v1.3 but hardwires v1.2

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#56378]
tldr: `go.mod` contains contradictory and invalid information, and endangers future dep upgrades.

See section "what we can do about this" below for suggested courses of action.

### Description of the problem

The `go.mod` does something illegal: it announces that the vendor tree provides gogoproto v1.3:

```
...
        github.com/gogo/protobuf v1.3.1
...
```

However it also pins the source code to our own fork which is anchored at v1.2:

```
replace github.com/gogo/protobuf => github.com/cockroachdb/gogoproto v1.2.1-0.20190102194534-ca10b809dba0
```

This contradicting specification creates a looming dependency breakage: if any package is upgraded to start depending on gogoproto  features specific to v1.3, then `go mod` will silently pick it up, and then the build will fail -- for example, with this error:

```
    variable or constant definition not found: proto.GoGoPackageIsVersion3
``` 

### Background (why this situation came to be)

CockroachDB was historically anchored to gogoproto v1.2. This is because when v1.2 came about, there was a bug in its implementation of generated `.Clone()` and `.Equal()` methods. So we forked the gogoproto repository at v1.2.4, and added a patch to this bug there. See [this upstream PR for details](https://github.com/gogo/protobuf/pull/529) 

Now, this particular bug was fixed in gogoproto v1.3, however since we used a forked repo we did not notice.

However, at some point in the last year we migrated from using `dep` to `go mod` to manage dependencies. During that upgrade, we mistakenly let `go mod` pick up version v1.3 as the *announced* tag in the list of dependencies, while retaining the  `replace` directive that was pinning the source code v1.2.

In effect, this mistake makes our `go.mod` "lie" to every other dependency about which version of gogoproto is available.

So if any other dependency starts using v1.3 features (as they could!), `go mod` would be unable to detect the incompatibility and we would get build failures without explanation.

### Why this is hard to fix

In addition to fixing the bug in v1.2 which prevented us from upgrading, v1.3 has introduced numerous optimizations and enhancements.

In particular, it introduces a new optimization to "optimize allocations when decoding protobufs": the new `Marshal()` methods *encode the fields from largest to smallest*, as opposed to increasing order of protobuf ID as previously.

This is causing a new / different problem in CockroachDB: we were assuming that the protobuf encoding is stable and deterministic for a given input data structure across all versions. This is because we want to be able to compare KV replicas byte-by-byte to determine equivalence; see `pkg/kv/kvserver/below_raft_protos_test.go`. This assumption is CockroachDB-specific -- protobuf allows different encoding orders in its spec, so the gogoproto optimization is legitimate.

Therefore, we are now unable to upgrade from v1.2 to a "native" gogoproto v1.3 because of this reordering.

### What we can do about this

- rewind the `go.mod` announced dependency to v1.2 so it does not "lie" anymore
  - pros: conceptually simple
  - cons: unclear whether this is supported at all by `go mod`
  - cons: makes us unable to benefit from the other optimizations in gogoproto

- upgrade to "native" gogoproto v1.3
  - pros: simplifies dep management
  - cons: requires new logic in cockroachdb to tolerate this reordering (i.e. introduce a "slow path" in the replica divergence detection code, that decodes the content to determine structural equivalence)
  - cons: makes replica consistency checks generally slower whenever we add new fields to protobufs that are larger than other fields

- fork gogoproto v1.3 and remove the specific undesirable reordering optimization
  - pros: minimal changes to cockroachdb
  - pros: some other optimizations in v1.3 are still available
  - cons: still need to manage a custom fork (but this is no worse than the status quo)

**Comment by knz** [comment]
cc @nvanbenschoten @bdarnell  @tbg for the technical analysis. Can you check the "What we can do about this" section and provide your impressions / recommendations?
 
cc @lunevalex for prioritization

**Comment by knz** [comment]
I think the reordering problem only exists for a few structs that are marshalled "under" KV - mvcc stats, raft etc. 

So instead of disabling the gogoproto optimization entirely, we could add an *option* that can be enabled for certain structs, specifically the few structs of interest here. The remainder of cockroachdb could use the native v1.3 encoding with the new order.

is my understanding correct?

**Comment by bdarnell** [comment]
Yes, that's correct. The affected structs rarely change, so we could even consider a manual solution (writing a custom marshal function using the old order for these functions and using it exclusively). 

> requires new logic in cockroachdb to tolerate this reordering (i.e. introduce a "slow path" in the replica divergence detection code, that decodes the content to determine structural equivalence)

My understanding is that the old format was simply sorted by tag number. It would then be fairly straightforward to write a CanonicalizeEncodedProto function that could be applied to the output of Marshal at *write* time of these structs (for less performance impact than an order-independent comparison function). This would be my preferred solution (as long as my assumptions are accurate).

**Comment by nvanbenschoten** [comment]
Thanks for writing this up @knz.

> My understanding is that the old format was simply sorted by tag number.

Yes, that's correct.

>> requires new logic in cockroachdb to tolerate this reordering (i.e. introduce a "slow path" in the replica divergence detection code, that decodes the content to determine structural equivalence)

> It would then be fairly straightforward to write a CanonicalizeEncodedProto function that could be applied to the output of Marshal at write time of these structs (for less performance impact than an order-independent comparison function). This would be my preferred solution (as long as my assumptions are accurate).

I'm not a huge fan of either of these options, but I think I'd prefer the first alternative if I had to pick. The slow-path in the replica divergence code wouldn't require an order-independent comparison function, but it would still likely slow down replica consistency checks. It would lead to unmarshalling and re-marshaling of `MVCCMetadata` objects (intents and inline values) in `Replica.sha512` to ensure byte-level equality of these values.

That would be relatively complex, as we'd have to determine which values this applies to (just `MVCCMetadata`?) and then maintain two marshaling functions for them. Or maybe we could generalize this through some kind of `CanonicalizeEncodedProto` function, like Ben suggests. However, I still like it better than the second option because at least it would be well isolated and it would not be in the hot path of performance-critical code. Additionally, it might be something that we could eventually remove through a long-running migration.

**Comment by knz** [comment]
1) One thing i am a bit uncomfortable about with the "slow path" is that a lot of data does not get rewritten frequently, so we'd get the slow path to kick in for a lot of ranges in the common path after a version upgrade. Maybe worth benchmarking the difference?

2) I don't understand your comment about "I still like it better than the second option because at least it would be well isolated". Ben's idea was to re-order the bytes when marshalling the 3 protobufs "below raft". We know more or less exactly what those are, so I find them quite well contained. What made you think otherwise?

**Comment by nvanbenschoten** [comment]
> One thing i am a bit uncomfortable about with the "slow path" is that a lot of data does not get rewritten frequently, so we'd get the slow path to kick in for a lot of ranges in the common path after a version upgrade. Maybe worth benchmarking the difference?

Benchmark the difference is a good idea. We already unmarshall `MVCCMetadata` structures in `Replica.sha512`, so it's possible that it wouldn't even be much slower. And it's worth pointing out again how much more latency-sensitive Raft application ("below Raft processing") is than replica consistency checks. It would be better to double to latency of replica consistency checks than to slow down Raft application by 5%.

> I don't understand your comment about "I still like it better than the second option because at least it would be well isolated". Ben's idea was to re-order the bytes when marshalling the 3 protobufs "below raft". We know more or less exactly what those are, so I find them quite well contained. What made you think otherwise?

There are 3 protobuf types that are marshaled below Raft, but I'm not sure that they are only marshaled in 3 places. For instance, MVCCMetadata wraps a number of different below-Raft writes to different keys. So we'd have to track to where all of these marshaling locations are and adapt them.

In addition to these two considerations, we should also focus on trying to introduce compatibility code that we can eventually remove, especially now that we're nearing the completion of a fancy new long-running migrations framework. It would be a shame if this issue resulted in subtle code deep in the guts of Raft that we will never be able to remove going forward.

**Comment by tbg** [comment]
Can someone post the PRs introducing the reordering? I'm aware of https://github.com/gogo/protobuf/pull/560/files but it's not entirely clear what the order change is. From a cursory look, is it "just" going through the fields in backwards order? i.e.

```
message Foo {
  int x = 1;
 int y = 2;
}
```

would first put `y` on the wire, then `x`, where as it was previously x followed by y? If it's "just" that, can't we rearrange the fields in our `.proto` files (atomically with the dep bump) so that the end result is the same?

**Comment by ajwerner** [comment]
This seems to be blocking https://github.com/cockroachdb/cockroach/issues/56208 which is, well, a bummer.

**Comment by knz** [comment]
I also have #56147 still out and possibly ready to merge, but it needs proper scrutiny by @tbg.

**Comment by tbg** [comment]
I don't understand #56147 - this PR would make sense to me if we didn't have a reordering problem, but don't we still have one?

**Comment by tbg** [comment]
@erikgrinaker mentioned in https://github.com/cockroachdb/cockroach/issues/56718#issuecomment-788038054 that [field presence](https://github.com/protocolbuffers/protobuf/blob/master/docs/implementing_proto3_presence.md) could help us bridge the gap to gogoproto (recall that one of our main gripes is that vanilla proto forces us to use tons of pointers).
Unfortunately, I think field presence is the opposite of what we want. Field presence allows making primitive types truly optional, i.e. it's really just turning `int`s into `*int`s as you can see here: https://github.com/tbg/goplay/blob/master/proto3presence/test.pb.go#L80-L84 (`IOpt` and `AOpt` use the new `optional` keyword). We'd want the opposite, elide the pointer and map `nil` to the zero value instead.

**Comment by andreimatei** [comment]
I'm a bit confused about where we left this off, and unfortunately I find myself caring. This issue started with a belief that gogoproto 1.3 changes the wire format of protos (in particular, our "below Raft" protos), which would create real trouble for us.  And we've blamed this serialization change on https://github.com/gogo/protobuf/pull/560 (although I'm not sure how we got to that particular PR).  @knz seemed to believe that the serialization changed to take into consideration the size of the fields, whereas the linked PR only deals with field numbers. In [this Slack discussion](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1604336170406900), we've doubted whether the serialization actually changed, or whether it just looks to our tests that it did: the linked PR contains [this change](https://github.com/gogo/protobuf/pull/560/commits/59e926049c56eeecfa66fcb0d1f1923355cf3336), which modifies the random proto generation function that our tests use. So it seems possible that everything is fine and we can upgrade gogoproto. @knz you've said that you'll investigate; I'm wondering if you ever did (not to shame you if you didn't... just checking).

**Comment by andreimatei** [comment]
To be more explicit, my hope is that `TestBelowRaftProtos` only *thinks* that the serialization has changed because the randomized constructor used by the test below has changed
https://github.com/cockroachdb/cockroach/blob/0668efbf891e36e77e14d24cf5837fac0c51a5f1/pkg/kv/kvserver/below_raft_protos_test.go#L164

**Comment by andreimatei** [comment]
FWIW, gogo/protobuf#560 explicitly claims to not change the serialization format in [this comment](https://github.com/gogo/protobuf/pull/560#issuecomment-482634142).

**Comment by knz** [comment]
I'm sorry I dropped the ball on this one; there was just a mounting list of things with a higher priority.
FWIW we've elevated this issue as a roadmap item in the v21.2 cycle.

I'm glad you were able to find out that the serialization format did not, in fact, change. What is the right way to reset the randomized constructor in the test so that the test passes?

**Comment by knz** [comment]
This PR:  #56147  is still open and if we can coerce the affected test to pass, I can rebase it and merge it.

**Comment by andreimatei** [comment]
I've verified that the encoding does not change by taking a `kvserverpb.RaftCommand` proto generated by a random test and marshaling it with and without #56147; the bytes are the same. If you look at the auto-generated randomized constructors for protos used by `TestBelowRaftProto` (e.g. `NewPopulatedInternalTimeSeriesData`), you see that the randomization has changed between the different gogoproto versions - various branches used to be taken with a 1/10 probability, and in the newer version they're taken with a 1/5 probability - which explains why `TestBelowRaftProtos` fails even though it fixes the random seed.

So I think we should go ahead with #56147. I've left some unrelated comments there, though.

> What is the right way to reset the randomized constructor in the test so that the test passes?

I think exactly what I believe you've done in the PR - rewrite the golden hashes expected by the test to match the ones the failures are reporting.
